lib/eig_trlm_3d.cpp

```diff
@@ -0,0 +1,701 @@
+    for (int t = 0; t < ortho_dim_size; t++) alpha_j[t] = active_3D[t] ? -alpha_j[t] : 0.0;
+    blas3d::axpby(alpha_j, v[j], unit, r[0]);
+
+    // r[t] = r[t] - b_{j-1}[t] * v_{j-1}[t]
+    // Only orthogonalise active problems
+    for (int t = 0; t < ortho_dim_size; t++) {
+      if (active_3D[t]) {
+        int start = (j > num_keep_3D[t]) ? j - 1 : 0;
+        if (j - start > 0) {
+
+          // Ensure we have enough 3D vectors
+          if ((int)vecs_t.size() < j - start) resize(vecs_t, j - start, csParamClone);
+
+          // Copy the 3D data into the 3D vectors, create beta array, and create
+          // pointers to the 3D vectors
+          std::vector<double> beta_t;
+          beta_t.reserve(j - start);
+          // Copy residual
+          blas3d::copy(t, blas3d::COPY_TO_3D, r_t, r[0]);
+          // Copy vectors
+          for (int i = start; i < j; i++) {
+            blas3d::copy(t, blas3d::COPY_TO_3D, vecs_t[i - start], v[i]);
+            beta_t.push_back(-beta_3D[t][i]);
+          }
+
+          // r[t] = r[t] - beta[t]{j-1} * v[t]{j-1}
+          blas::block::axpy(beta_t, {vecs_t.begin(), vecs_t.begin() + j - start}, r_t);
+
+          // Copy residual back to 4D vector
+          blas3d::copy(t, blas3d::COPY_FROM_3D, r_t, r[0]);
+        }
+      }
+    }
+
+    // Orthogonalise r against the Krylov space
+    for (int k = 0; k < 1; k++) blockOrthogonalize3D(v, r, j + 1);    //  future work: up to 4 times???
+
+    // b_j[t] = ||r[t]||
+    blas3d::reDotProduct(beta_j, r[0], r[0]);
+    for (int t = 0; t < ortho_dim_size; t++) beta_j[t] = active_3D[t] ? sqrt(beta_j[t]) : 0.0;
+
+    // Prepare next step.
+    // v_{j+1}[t] = r[t] / b_{j}[t]
+    for (int t = 0; t < ortho_dim_size; t++) {
+      if (active_3D[t]) {
+        beta_3D[t][j] = beta_j[t];
+        beta_j[t] = 1.0 / beta_j[t];
+      }
+    }
+
+    std::vector<double> c(ortho_dim_size);
+    for (int t = 0; t < ortho_dim_size; t++) c[t] = beta_j[t] == 0.0 ? 1.0 : 0.0;
+    blas3d::axpby(beta_j, r[0], c, v[j + 1]);
+  }
+
+  void TRLM3D::reorder3D(std::vector<ColorSpinorField> &kSpace)
+  {
+    for (int t = 0; t < ortho_dim_size; t++) {
+      int i = 0;
+      if (reverse) {
+        while (i < n_kr) {
+          if ((i == 0) || (alpha_3D[t][i - 1] >= alpha_3D[t][i]))
+            i++;
+          else {
+            std::swap(alpha_3D[t][i], alpha_3D[t][i - 1]);
+            blas3d::swap(t, kSpace[i], kSpace[i - 1]);
+            i--;
+          }
+        }
+      } else {
+        while (i < n_kr) {
+          if ((i == 0) || (alpha_3D[t][i - 1] <= alpha_3D[t][i]))
+            i++;
+          else {
+            std::swap(alpha_3D[t][i], alpha_3D[t][i - 1]);
+            blas3d::swap(t, kSpace[i], kSpace[i - 1]);
+            i--;
+          }
+        }
+      }
+    }
+  }
+
+
+  void TRLM3D::eigensolveFromArrowMat3D()
+  {
+    getProfile().TPSTART(QUDA_PROFILE_EIGEN);
+
+    // Loop over the 3D problems
+#ifdef _OPENMP
```

Comment:
Is there a particular need for this `#ifdef` or can we just get rid of it?

Reply:
Indeed the guard isn't needed, will be nuked.

Reply:
c5b568171